packages/web3-common/src/errors.ts

```diff
@@ -0,0 +1,333 @@
+
+export abstract class Web3Error extends Error {
+	public readonly name: string;
+	public readonly abstract code: number;
+
+	public constructor(msg: string) {
+		super(msg);
+		this.name = this.constructor.name;
+		Error.captureStackTrace(this, Web3Error);
+	}
+
+	public toJSON() {
+		return { name: this.name, code: this.code, message: this.message };
+	}
+}
+
+export class InvalidNumberOfParamsError extends Web3Error {
+	public code = ERR_PARAM;
+
+	public constructor(public got: number, public expected: number, public method: string) {
+		super(`Invalid number of parameters for "${method}". Got "${got}" expected "${expected}"!`);
+	}
+
+	public toJSON() {
+		return {
+			...super.toJSON(),
+			got: this.got,
+			expected: this.expected,
+			method: this.method,
+		};
+	}
+}
+
+export class ConnectionError extends Web3Error {
+	public code = ERR_CONN;
+	public errorCode?: string;
+	public errorReason?: string;
+
+	public constructor(message: string, event?: ConnectionEvent) {
+		super(message);
+
+		if (event) {
+			this.errorCode = event.code;
+			this.errorReason = event.reason;
+		}
+	}
+
+	public toJSON() {
+		return { ...super.toJSON(), errorCode: this.errorCode, errorReason: this.errorReason };
+	}
+}
+
+export class InvalidConnectionError extends ConnectionError {
+	public constructor(public host: string, event?: ConnectionEvent) {
+		super(`CONNECTION ERROR: Couldn't connect to node ${host}.`, event);
+		this.code = ERR_CONN_INVALID;
+	}
+
+	public toJSON() {
+		return { ...super.toJSON(), host: this.host };
+	}
+}
+
+export class ConnectionTimeoutError extends ConnectionError {
+	public constructor(public duration: number) {
+		super(`CONNECTION TIMEOUT: timeout of ${duration}ms achieved`);
+		this.code = ERR_CONN_TIMEOUT;
+	}
+
+	public toJSON() {
+		return { ...super.toJSON(), duration: this.duration };
+	}
+}
+
+export class ConnectionNotOpenError extends ConnectionError {
+	public constructor(event?: ConnectionEvent) {
+		super('Connection not open on send()', event);
+		this.code = ERR_CONN_NOT_OPEN;
+	}
+}
+
+export class ConnectionCloseError extends ConnectionError {
+	public constructor(event?: ConnectionEvent) {
+		super(
+			`CONNECTION ERROR: The connection got closed with the close code ${
+				event?.code ?? ''
+			} and the following reason string ${event?.reason ?? ''}`,
+			event,
+		);
+		this.code = ERR_CONN_CLOSE;
+	}
+}
+
+export class MaxAttemptsReachedOnReconnectingError extends ConnectionError {
+	public constructor() {
+		super('Maximum number of reconnect attempts reached!');
+		this.code = ERR_CONN_MAX_ATTEMPTS;
+	}
+}
+
+export class PendingRequestsOnReconnectingError extends ConnectionError {
+	public constructor() {
+		super('CONNECTION ERROR: Provider started to reconnect before the response got received!');
+		this.code = ERR_CONN_PENDING_REQUESTS;
+	}
+}
+
+export class InvalidProviderError extends Web3Error {
+	public code = ERR_INVALID_PROVIDER;
+
+	public constructor() {
```

Comment:
I think this error refers to what should be called the `Client` based off the definitions defined [here](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#definitions). As we are constructing the `Provider` objects via class instantiation, I don't think we could have an invalid one? But perhaps there's a use case where a user provides their own instantiated `Provider`, and we throw this when their `Provider` doesn't match our defined interface?

Regardless, we're missing a `InvalidClientError` that should take the provided `clientUrl` as a constructor argument to be used in the error message

Reply:
For the compiled code running in JS we can't enforce that every provider must be inherited by our abstract class interface. So would be nice to do also validate on runtime.

Reply:
> we throw this when their Provider doesn't match our defined interface

I mean we would check that `request`, `setProvider`, etc. methods are available on whatever Provider they pass. If we did this, then we'd have separate errors for invalid providers and clients e.g.

```typescript
if (userProvider.request === undefined) throw InvalidProviderError()
```

My suggestion is to revert the `InvalidProviderError` to what you originally had since it exists in `1.x` (although I'm not entirely sure where we are going to use it), and create a new error called `InvalidClientError` that takes the `clientUrl`. `InvalidClientError` will be used when performing validation on the URL provided by the user when instantiating `Web3()` , `Web3.httpProvider()`, etc.

Reply:
I am also not sure where and how these errors are gonna used. So for now I just created what we already had. I suggest we move forward with it and as its a work in progress, if we ever need more particular error objects we can create.